storage: don't panic on unknown (batch) request

tbg opened the pull request:
When a new BatchRequest type is introduced, we have to be careful to
not send it to older versions of CockroachDB because the new request
will simply be dropped when unmarshaling the proto. This leads to a
`ba.Requests` slice containing zero values, which leads to panics.

This commit adds a guard against unknown requests into `(*Node).Batch`
so that beginning in v2.0, it is less dramatic to forget the correct
cluster migration which gates use of new request types properly.
(I am not suggesting leaving out these migrations when this commit
has matured).

See:
https://github.com/cockroachdb/cockroach/pull/21345#issuecomment-357986640.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21466)
<!-- Reviewable:end -->

tschottdorf commented:
NB: I'm planning to cherry-pick this to v1.1., though I can be convinced otherwise.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

I'm -1 on cherry-picking to 1.1. If we make sure we have migrations for everything in 2.0, it's unnecessary. We currently allow upgrading from any 1.1.x release to 2.0.y; I'd rather not introduce a requirement that the cluster be upgraded to at least 1.1.5 before 2.0 unless we have to. 

---

Reviewed 4 of 4 files at r1.
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/acceptance/version_upgrade_test.go, line 70 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21466#-L2zd1FKIr73PAyWkqNF:-L2zd1FKIr73PAyWkqNG:byipi08) ([raw file](https://github.com/cockroachdb/cockroach/blob/823d8a89d87f949f0216dcb06a711675f463a3d4/pkg/acceptance/version_upgrade_test.go#L70)):*
> ```Go
> 	// The answer is "not well". v1.1 has the same problem except in late patch
> 	// releases. v2.0 does not have this issue, but it should be tested once (if)
> 	// we introduce a new command to v2.1 by adapting the below code. See
> ```

So the idea behind this block is that the first new command in 2.1 should uncomment the ba.Add line with the new command to provide a test that it doesn't crash the old node?

---

*[pkg/server/node.go, line 835 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21466#-L2zdJj06LCO79_zYVhm:-L2zdJj06LCO79_zYVhn:b4rqqrg) ([raw file](https://github.com/cockroachdb/cockroach/blob/823d8a89d87f949f0216dcb06a711675f463a3d4/pkg/server/node.go#L835)):*
> ```Go
> 	for _, req := range reqs {
> 		if req.GetValue() == nil {
> 			return log.Safe("unknown request in batch")
> ```

I think it would be best if this was a roachpb.Error so we can cleanly catch it on the client side (for example with RecomputeStats, we could simply fail silently whenever we get the old-version error)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21466#-:-L2zdY-O7vb8VHV6mxcW:b-804crn)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: don't panic on unknown (batch) request

When a new BatchRequest type is introduced, we have to be careful to
not send it to older versions of CockroachDB because the new request
will simply be dropped when unmarshaling the proto. This leads to a
`ba.Requests` slice containing zero values, which leads to panics.

This commit adds a guard against unknown requests into `(*Node).Batch`
so that beginning in v2.0, it is less dramatic to forget the correct
cluster migration which gates use of new request types properly.
(I am not suggesting leaving out these migrations when this commit
has matured).

See:
https://github.com/cockroachdb/cockroach/pull/21345#issuecomment-357986640.

Release note: None

bdarnell commented:
LGTM

---

Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21466#-:-L2zrnr5msbWu6t7nflm:bb6ywmx)*
<!-- Sent from Reviewable.io -->
